File: DataStructure/src/lineartable/LRUCache.java
package lineartable;

import java.util.HashMap;
import java.util.Map;

public class LRUCache {
    Map<Integer, DNode> map ;
    DNode lhead;
    DNode ltail;
    class DNode{
        int value;
        DNode prev;
        DNode next;
        private DNode(int val){
            value = val;
            prev = null;
            next = null;
        }
    }
    int capacity = 0;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        map = new HashMap<Integer, DNode>();
        lhead = new DNode(-1);
        ltail = new DNode(-1);
        lhead.next = ltail;
        ltail.prev = lhead;
    }

    public int get(int key) {
        if(map.get(key) == null){
            return -1;
        }else{
            int val = map.get(key).value;
            updateRecentList(key,val);
            return val;
        }
    }

    public void put(int key, int value) {
        if(map.size() < capacity){
            DNode node = new DNode(value);
            map.put(key,node);
            updateRecentList(key,value);
        }else if(map.size() == capacity){
            DNode node = map.get(key);
            updateRecentList(key,value);
        }else{
            System.out.println("Exceed capacity!");
        }
    }
    private void remove(int key){
        DNode node = map.get(key);
        if( node == null) return;
        else{
            removeNode(node);
            map.remove(key);
        }
    }
    private void updateRecentList(int key,int value){
        if(map.get(key) == null){
            DNode node = new DNode(value);
            DNode newnode = addFirst(node);
            map.put(key,newnode);
        }else{
            DNode node = map.get(key);
            removeNode(node);
            DNode newnode = addFirst(node);
            map.put(key,newnode);
        }
    }
    private DNode addFirst(DNode node){
        node.next = lhead.next;
        node.prev = lhead;
        lhead.next = node;
        return node;
    }
    private void removeNode(DNode node){
        node.prev.next = node.next;
        node.next.prev = node.prev;
        node = null;
    }

    public static void main(String[] args) {
        String s = s.substring()
    }
}

File: DataStructure/src/lineartable/LRUCache.cs
public class LRUCache{
    private Dictionary<int, DNode> _map;
    private DNode _dummyHead;
    private Node _dummyTail;
    private class DNode{ // double LinkedList
        int value;
        DNode prev;
        DNode next;
        private DNode(int val){
            value = val;
            prev = null;
            next = null;
        }
    }
    public int capacity;
    private int size;

    public LRUCache(int defaultSize) {
        this.capacity = defaultSize;
        _size = 0;
        _map = new Dictionary<int, DNode>();
        _dummyHead = new DNode(-1);
        _dummyTail = new DNode(-1);
        _dummyHead.next = _dummyTail;
        _dummyTail.prev = _dummyHead;
    }

    public int get(int key) {
        if(_map.ContainsKey(key)){
            DNode node = _map[key];
            removeNode(node);
            addToFirst(node);
            return node.value;
        }
        return -1;
    }

    public void set(int key, int value) {
        if(_map.ContainsKey(key)){
            DNode node = _map[key];
            node.value = value;
            removeNode(node);
            addToFirst(node);
        }else{
            if(_size == capacity){
                map.Remove(_dummyTail.prev.key);
                removeNode(_dummyTail.prev);
                _size--;
            }
            DNode newNode = new DNode(key, value);
            _map.Add(key,newNode);
            addToFirst(newNode);
            _size++;
        }
    }

    private void addToFirst(DNode node){
        node.next = _dummyHead.next;
        node.prev = _dummyHead;
        _dummyHead.next.prev = node;
        _dummyHead.next = node;
    }
    private void removeNode(DNode node){
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }
}